daft/execution/physical_plan.py

```diff
@@ -1462,34 +1463,28 @@ def materialize(
 ) -> MaterializedPhysicalPlan:
     """Materialize the child plan.
 
-    Repeatedly yields either a PartitionTask (to produce an intermediate partition)
-    or a PartitionT (which is part of the final result).
+    Repeatedly yields either a PartitionTask (to produce an intermediate partition) and
+    a boolean indicating whether or not this is a final result.
+
+    Args:
+        child_plan: Child "in progress" plan
     """
 
-    materializations: deque[SingleOutputPartitionTask[PartitionT]] = deque()
     stage_id = next(stage_id_counter)
-    while True:
-        # Check if any inputs finished executing.
-        while len(materializations) > 0 and materializations[0].done():
-            done_task = materializations.popleft()
-            yield done_task.result()
 
-        # Materialize a single dependency.
-        try:
-            step = next(child_plan)
-            if isinstance(step, PartitionTaskBuilder):
-                step = step.finalize_partition_task_single_output(stage_id=stage_id)
-                materializations.append(step)
-            assert isinstance(step, (PartitionTask, type(None)))
-
-            yield step
-
-        except StopIteration:
-            if len(materializations) > 0:
-                logger.debug("materialize blocked on completion of all sources: %s", materializations)
-                yield None
-            else:
-                return
+    # Materialize a single dependency.
+    for step in child_plan:
+        # None indicates that no forward progress can be made yet
+        if step is None:
+            yield (step, False)
+        # PartitionTaskBuilder indicates that this is a final materialization
+        elif isinstance(step, PartitionTaskBuilder):
+            step = step.finalize_partition_task_single_output(stage_id=stage_id)
+            yield (step, True)
+        # PartitionTask indicates an intermediate result
+        else:
+            assert isinstance(step, PartitionTask)
+            yield (step, False)
```

Comment:
This is the major change to our `MaterializedPhysicalPlan`. It no longer keeps track of `materializations` and yields `MaterializedResults`. Instead, it naively yields `PartitionTasks` and the responsibility of keeping track of these materializations is lifted to the caller.